### ISSUE TITLE
strange widget alignment behavior

### ISSUE THREAD

**Issue by bill10** [bokeh/bokeh#4917]
A picture is worth more than a thousand words. Please see the figure attached below (the clash between widgets). Thank you. 

The code is as the following, running in bokeh server.

```
symbol_box=TextInput(value="Symbol", title=None, width=100)
button = Button(label="GO", button_type='success', width=60)
plot = figure(width=900, height=300, x_axis_type="date time"
curdoc().add_root(column([row([symbol_box,button]), plot], sizing_mode='scale_width'))

```

![screen shot 2016-08-02 at 4 18 37 pm](https://cloud.githubusercontent.com/assets/4839264/17321938/97c06d36-58cd-11e6-9c24-94166c03fbba.png)

**Comment by tobyhodges** [comment]
I have been able to reproduce this. I found a workaround for the overlap, by wrapping both widgets in `widgetbox` layout objects. But this still leaves a strange alignment issue, where the button floats higher than the textinput box.

```Python
from bokeh.models import TextInput, Button
from bokeh.plotting import figure, curdoc
from bokeh.layouts import row, column, widgetbox

symbol_box=TextInput(value="Symbol", title=None, width=100)
button = Button(label="GO", button_type='success', width=60)
plot = figure(width=900, height=300, x_axis_type="datetime")
curdoc().add_root(column(row(widgetbox(symbol_box), widgetbox(button)), plot, sizing_mode='scale_width'))
```

![screen shot 2016-08-02 at 16 43 59](https://cloud.githubusercontent.com/assets/9694524/17332809/674da8a0-58d0-11e6-8407-3548a95d25c0.png)

**Comment by bill10** [comment]
@tobyhodges thanks for the help. The horizontal alignment issue is due to the fact that InputBox has a "reserved spot" for Title, I think.

**Comment by birdsarah** [comment]
Currently the support for laying out widgets vertically is much better than the support for laying out widgets horizontally. Definitely did not test this combination of a text input and button horizontally with the text input label being empty.

**Comment by ongietorri** [comment]
Same issue here in 0.12.11 with a minimal example:

```
from bokeh.io import output_notebook, show
from bokeh.layouts import widgetbox, row
from bokeh.models.widgets import Button, TextInput

output_notebook()

# create some widgets
button0 = Button(label="Button 0: ok", width=100)
button1 = Button(label="Button 1: ok", width=100)
button2 = TextInput(value="TextInput 2: V-ALIGN FAIL", width=100)

show(row(widgetbox(button0, width=130), widgetbox(button1, width=130), widgetbox(button2)))
```

![image](https://user-images.githubusercontent.com/5689266/33482457-6ff74cc0-d699-11e7-9149-778c7835c0d6.png)

**Comment by iaindinw** [comment]
Just to add to this issue, it's still present in version 0.12.13, it's not solely attributable to Buttons and their lack of title labels. You can get similar weird horizontal alignment issues with Select and DatePicker objects, for example:

`from bokeh.io import output_notebook, show
from bokeh.layouts import widgetbox, row, layout
from bokeh.models.widgets import Select, DatePicker
from datetime import date

date_select = DatePicker(max_date=date.today(),
                          value=date(2018, 1, 1), title='Pick Date')

value_select = Select(title='Pick Value', options=list(['1', '2', '3']),
                      value='1')
lay = layout(row([date_select, value_select]))

show(lay)`

Results in:

![image](https://user-images.githubusercontent.com/18350007/35504983-873d1880-04dc-11e8-83d7-8be6ab26469f.png)

**Comment by omnesia0rg** [comment]
The issue persists in version 0.13.0
I believe this has to do with the fact that some widgets have titles, even if the text is empty. This pushes the the actual widget down.
One tedious workaround is to wrap the non-titled widget (e.g. `Button`) in a `Column()` together with a `Div()` to push it down as well.

Cases:
```
from bokeh.layouts import Column, Row, WidgetBox
from bokeh.models.widgets import Select, Button, Div
from bokeh.io import show, output_notebook
output_notebook()

w=200

initial = Row(
    Select(title="Some Title", value="Select", options=["Select"],width=w),
    Button(label="Button", width=w)
)

add_label = Row(
    Select(title="Some Title", value="Select", options=["Select"],width=w),
    Column(
        Div(text="", height=2), # Height differs when the Select has an empty title
        Button(label="ButtonWithDiv", width=w)
    )
)

fix_overlap = Row(
    WidgetBox(
        Select(title="Some Title", value="SelectInBox", options=["SelectInBox"],width=w),
        width=w+20 # NOTICE the larger width
    ),
    Column(
        Div(text="", height=2),
        Button(label="ButtonWithDiv", width=w)
    )
)

no_title = Row(
    Select(title="", value="SelectNoTitle", options=["SelectNoTitle"],width=w),
    Button(label="Button", width=w)
)


fix_overlap_no_title = Row(
    WidgetBox(
        Select(title="", value="SelectNoTitle", options=["SelectNoTitle"],width=w),
        width=w+20 # NOTICE the larger width
    ),
    Column(
        Div(text="", height=1), # Notice smaller height
        Button(label="ButtonWithDiv", width=w)
    )
)

show(
    Column(
        initial,
        add_label,
        fix_overlap,
        no_title,
        fix_overlap_no_title
    )
)
```

Output:
![2018101111 27 22_screenshot_sel](https://user-images.githubusercontent.com/24209580/46795120-a4715e80-cd49-11e8-85c0-c271a15c3de5.png)

**Comment by JEGBergqvist** [comment]
Is this related to the issue of empty (i.e. =None) label allocates empty space over text input widgets?

**Comment by JEGBergqvist** [comment]
I am thinking this is perhaps higher prio than one might think.It seems I am not alone in spending time for trying to workaround empty (including =None) titles for text input boxes taking up space above the actual box. The result is a wide range of tricks and ad hoc adjustments on the web (ranging from various JS / CSS patches to DOM Layout wrappers, Spacer-solutions, etc)

Also, Bokeh is such a great idea, bad alignment in a GUI is a bit of a waste from a perception / impression point of view for the framework as such.

**Comment by bryevdv** [comment]
> I am thinking this is perhaps higher prio than one might think.

There is a large WIP Pr to address many layout related issues at once. But as you can imagine a complete re-work of layout is a large drawn out undertaking both in terms of implementation and the need for careful review and testing. 

> Also, Bokeh is such a great idea, bad alignment in a GUI is a bit of a waste from a perception / impression point of view for the framework as such.
 
We are doing the best we can with the resources available. If you have the skills/experience to help, it is certainly very welcome.

**Comment by bryevdv** [comment]
FYI the last example looks like this on 1.1.0dev9:

<img width="433" alt="screen shot 2019-03-07 at 12 47 38 pm" src="https://user-images.githubusercontent.com/1078448/53987971-44ec6480-40d7-11e9-85cb-b5159fb3fbc0.png">

**Comment by bryevdv** [comment]
also @tobyhodges code now looks like this:

<img width="1113" alt="screen shot 2019-03-07 at 12 50 14 pm" src="https://user-images.githubusercontent.com/1078448/53988127-9b59a300-40d7-11e9-8637-1db27eb254f7.png">

**Comment by bryevdv** [comment]
I think this issue has been sufficiently resolved by the 1.1 layout work, so I am closing. In the interest of having more focused easy to digest issues, please make new issues for anything follow-on
